tests/ops/functions/test_eigvals.py

```diff
@@ -0,0 +1,405 @@
+
+        assert np.allclose(np.sort(res.real), np.sort(expected.real))
+        assert np.allclose(np.sort(res.imag), np.sort(expected.imag))
+
+    def test_multiple_operations_tape(self):
+        """Check the eigenvalues for a tape containing multiple gates"""
+
+        with qml.tape.QuantumTape() as tape:
+            qml.PauliX(wires="a")
+            qml.S(wires="b")
+            qml.Hadamard(wires="c")
+            qml.CNOT(wires=["b", "c"])
+
+        with pytest.warns(UserWarning, match="the eigenvalues will be computed numerically"):
+            res = qml.eigvals(tape)
+
+        expected = np.linalg.eigvals(np.kron(I, CNOT) @ np.kron(X, np.kron(S, H)))
+        assert np.allclose(res, expected)
+
+    def test_multiple_operations_qfunc(self):
+        """Check the eigenvalues for a qfunc containing multiple gates"""
+
+        def testcircuit():
+            qml.PauliX(wires="a")
+            qml.S(wires="b")
+            qml.Hadamard(wires="c")
+            qml.CNOT(wires=["b", "c"])
+
+        with pytest.warns(UserWarning, match="the eigenvalues will be computed numerically"):
+            res = qml.eigvals(testcircuit)()
+
+        expected = np.linalg.eigvals(np.kron(I, CNOT) @ np.kron(X, np.kron(S, H)))
+        assert np.allclose(res, expected)
+
+    def test_multiple_operations_qnode(self):
+        """Check the eigenvalues for a QNode containing multiple gates"""
+        dev = qml.device("default.qubit", wires=["a", "b", "c"])
+
+        @qml.qnode(dev)
+        def testcircuit():
+            qml.PauliX(wires="a")
+            qml.adjoint(qml.S)(wires="b")
+            qml.Hadamard(wires="c")
+            qml.CNOT(wires=["b", "c"])
+            return qml.expval(qml.PauliZ("a"))
+
+        with pytest.warns(UserWarning, match="the eigenvalues will be computed numerically"):
+            res = qml.eigvals(testcircuit)()
+
+        expected = np.linalg.eigvals(np.kron(I, CNOT) @ np.kron(X, np.kron(np.linalg.inv(S), H)))
+        assert np.allclose(res, expected)
+
+
+class TestTemplates:
+    """These tests are useful as they test operators that might not have
+    matrix forms defined, requiring decomposition."""
+
+    def test_instantiated(self):
+        """Test an instantiated template"""
+        weights = np.array([[[0.1, 0.2, 0.3], [0.1, 0.2, 0.3]]])
+        op = qml.StronglyEntanglingLayers(weights, wires=[0, 1])
+
+        with pytest.warns(UserWarning, match="the eigenvalues will be computed numerically"):
+            res = qml.eigvals(op)
+
+        with qml.tape.QuantumTape() as tape:
+            op.decomposition()
+
+        with pytest.warns(UserWarning, match="the eigenvalues will be computed numerically"):
+            expected = qml.eigvals(tape)
+
+        assert np.allclose(res, expected)
+
+    def test_qfunc(self):
+        """Test a template used within a qfunc"""
+
+        def circuit(weights, x):
+            qml.StronglyEntanglingLayers(weights, wires=[0, 1])

```

Comment:
No need to do anything here, but in future I'd find it better to define a new op that has a decomposition but no matrix representation, to avoid making tests dependent on specific subclasses we may change...

Reply:
Good point!